nilearn/glm/first_level/first_level.py

```diff
@@ -959,107 +950,387 @@ def first_level_from_bids(dataset_path, task_label, space_label=None,
+        )
+        events = get_bids_files(dataset_path,
+                                modality_folder='func',
+                                file_tag='events',
+                                file_type='tsv',
+                                sub_label=sub_label,
+                                filters=events_filters)
+        _check_bids_events_list(events=events,
+                                imgs=imgs,
+                                sub_label=sub_label,
+                                task_label=task_label,
+                                dataset_path=dataset_path,
+                                events_filters=events_filters,)
+
+        if verbose:
+            print('Found the following events files',
+                  f'for subject {sub_label}',
+                  f'for filter: {filters}:\n',
+                  f'{events}\n')
+
+        events = [pd.read_csv(event, sep='\t', index_col=None)
+                  for event in events]
+
+        models_events.append(events)
+
+        # Get confounds.
+        # If not found it will be assumed there are none.
         # If there are confounds, they are assumed to be present for all runs.
-        confounds = get_bids_files(derivatives_path, modality_folder='func',
+        confounds_filters = _supported_bids_filter()["raw"] + \
+                            _supported_bids_filter()["derivatives"]
+        confounds_filters.remove('desc')
+        filters = _filter_for_bids_query(task_label=task_label,
+                                         supported_filters=confounds_filters,
+                                         extra_filter=img_filters)
+        confounds = get_bids_files(derivatives_path,
+                                   modality_folder='func',
                                    file_tag='desc-confounds*',
-                                   file_type='tsv', sub_label=sub_label,
+                                   file_type='tsv',
+                                   sub_label=sub_label,
                                    filters=filters)
 
         if confounds:
             if len(confounds) != len(imgs):
-                raise ValueError('%d confounds.tsv files found for %d bold '
-                                 'files. Same number of confound files as '
-                                 'the number of runs is expected' %
-                                 (len(events), len(imgs)))
+                raise ValueError(f"{len(events)} confounds.tsv files found "
+                                 f"for {len(imgs)} bold files. "
+                                 "Same number of confound files as "
+                                 "the number of runs is expected")
             confounds = [pd.read_csv(c, sep='\t', index_col=None)
                          for c in confounds]
+
             models_confounds.append(confounds)
 
     return models, models_run_imgs, models_events, models_confounds
+
+
+def _supported_bids_filter() -> dict[str, list[str]]:
```

Comment:
This is pretty much a duplication of the `_bids_entities` function in data_gen.
It may be useful to start creating a module with bids related utility functions.